src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -393,24 +401,34 @@ public override bool TryWrite(T item)
                     {
                         // The channel is full.  Just ignore the item being added
                         // but say we added it.
+                        Monitor.Exit(parent.SyncObj);
+                        releaseLock = false;
+                        parent._itemDropped?.Invoke(item);
                         return true;
                     }
                     else
                     {
                         // The channel is full, and we're in a dropping mode.
-                        // Drop either the oldest or the newest and write the new item.
-                        if (parent._mode == BoundedChannelFullMode.DropNewest)
-                        {
-                            parent._items.DequeueTail();
-                        }
-                        else
-                        {
+                        // Drop either the oldest or the newest
+                        T droppedItem = parent._mode == BoundedChannelFullMode.DropNewest ?
+                            parent._items.DequeueTail() :
                             parent._items.DequeueHead();
-                        }
+
+                        Monitor.Exit(parent.SyncObj);
+                        releaseLock = false;
+                        parent._itemDropped?.Invoke(droppedItem);
+
                         parent._items.EnqueueTail(item);
```

Comment:
Enqueueing must be done while the lock is held.
```suggestion
                        parent._items.EnqueueTail(item);
                        Monitor.Exit(parent.SyncObj);                        
                        releaseLock = false;
                        parent._itemDropped?.Invoke(droppedItem);
```

Reply:
Done